src/AssocList.mo

```diff
@@ -82,38 +82,63 @@ module {
   ///
   /// *Runtime and space assumes that `equal` runs in O(1) time and space.
   public func replace<K, V>(
-    map : AssocList<K, V>,
-    key : K,
-    equal : (K, K) -> Bool,
-    value : ?V
-  ) : (AssocList<K, V>, ?V) {
-    var prev : ?V = null;
-    func rec(al : AssocList<K, V>) : AssocList<K, V> {
-      switch (al) {
-        case (?((hd_k, hd_v), tl)) {
-          if (equal(key, hd_k)) {
-            // if value is null, remove the key; otherwise, replace key's old value
-            // return old value
-            prev := ?hd_v;
-            switch value {
-              case (null) { tl };
-              case (?value) { ?((hd_k, value), tl) }
+      map : AssocList<K, V>,
+      key : K,
+      equal : (K, K) -> Bool,
+      value : ?V
+    ) : (AssocList<K, V>, ?V) {
+      var prev : ?V = null;
+      switch value {
+        case null {
+          func del(al : AssocList<K, V>) : AssocList<K, V> {
+            switch (al) {
+              case (?((hd_k, hd_v), tl)) {
+                if (equal(key, hd_k)) {
+                  prev := ?hd_v;
+                  tl
+                } else {
+                  let tl1 = del(tl);
+                  switch (prev) {
+                    case (?_)  { ?((hd_k, hd_v), tl1) };
+                    case (null) { al }
+                  }
+                }
+              };
+              case (null) {
+                null
+              }
+            }
+          };
+          (del(map), prev)
+      };
+      case (?value) {
+        func ins(al : AssocList<K, V>) : AssocList<K, V> {
```

Comment:
It looks to me that the helper function `ins` is identical to `del`? Not sure if I miss some relevant difference. Otherwise, it could be maybe unified to one helper function, to remove a potential existing previous value (both for deletion and first step before replacing.)

Reply:
Actually  you are right, I think its just deleting the old binding, returning original if possible. 

Need lunch though.